LightGBM/data.py

```diff
@@ -36,37 +36,94 @@ def downcast(df, verbose=False):
                 elif c_min > np.finfo(np.float32).min and c_max < np.finfo(np.float32).max:
                     df[col] = df[col].astype(np.float32)
                 else:
-                    df[col] = df[col].astype(np.float64)   
-        elif col == 'date':
-            df[col] = pd.to_datetime(df[col], format='%Y-%m-%d')
-        elif str(col_type) == 'object':
-            df[col] = df[col].astype('category')
-        else:
-            print(f'Unexpected dtype {col_type} in dataframe.')   
+                    df[col] = df[col].astype(np.float64)
 
     end_mem = df.memory_usage().sum() / 1024**2
     if verbose: 
         print('Mem. usage decreased to {:5.2f} Mb ({:.1f}% reduction)'.format(end_mem, 100 * (start_mem - end_mem) / start_mem))
     return df
 
 
-def cat_to_int(df, ignore=[]):
+def obj_as_cat_int(df, ignore=[]):
     """
-    Convert categorical columns to integer.
+    Convert object columns to categorical integers.
 
     Args:
         df = [pd.DataFrame] pandas dataframe
     
     Returns [pd.DataFrame]:
-        Data where categorical columns are encoded as integers.
+        Data where object columns are encoded as categorical integers.
     """
-    cat_cols = df.select_dtypes(include='category').columns
-    for col in cat_cols:
+    obj_cols = df.select_dtypes(include='object').columns
+    for col in obj_cols:
         if col not in ignore:
+            df[col] = df[col].astype('category')
             df[col] = df[col].cat.codes.astype("int16")
             df[col] -= df[col].min()
     return df
 
 
+def optimize_df(calendar, prices, sales, days=None, verbose=False):
+    """
+    Optimize dataframe.
+
+    Args:
+        calendar = [pd.DataFrame] dates of product sales
+        prices   = [pd.DataFrame] price of the products sold per store and date
+        sales    = [pd.DataFrame] historical daily unit sales data per product and store 
+        days     = [int] number of days to keep
+        verbose  = [boolean] if True, print memory reduction
+
+    Returns [(pd.DataFrame) * 3]
```

Comment:
`(pd.DataFrame) * 3` doesn't really make sense.

With a list: `[3]*4 == [3, 3, 3, 3]`
With a tuple: `(3,)*4 == (3, 3, 3, 3)`
Now: `(3)*4 == 12`

I use the list notation to save the comma character, but if you prefer the tuple syntax, then the comma is necessary

Reply:
I see, I changed it to the list notation everywhere to stay consistent with what you use.